src/poliastro/core/iod.py

```diff
@@ -111,42 +111,75 @@ def vallado(k, r0, r, tof, short, numiter, rtol):
 
     A = t_m * (norm_r * norm_r0 * (1 + cos_dnu)) ** 0.5
 
-    if A == 0.0:
+    if abs(A) <= 0.2:  # Not exactly zero
         raise RuntimeError("Cannot compute orbit, phase angle is 180 degrees")
 
-    psi = 0.0
-    psi_low = -4 * np.pi ** 2
-    psi_up = 4 * np.pi ** 2
+    psi = 0
+
+    if M == 0:  # Usual M == 0 case
+        psi_low = -4.0 * np.pi ** 2
+        psi_up = 4.0 * np.pi ** 2
+
+    else:
+        psi_low = 4.0 * M ** 2 * np.pi ** 2
+        psi_up = 4.0 * (M + 1.0) ** 2 * np.pi ** 2
+
+    psi = _initial_guess_vallado(psi, psi_low, psi_up, M, tof, direct)
 
     count = 0
+    ynegktr = 1
+    tof_new = -10
+
+    while count < numiter and ynegktr <= 10:
+
+        if abs(c2(psi) > rtol):
+            y = norm_r0_plus_norm_r + A * (1.0 - psi * c3(psi)) / c2(psi) ** 0.5
+        else:
+            y = norm_r0_plus_norm_r
 
-    while count < numiter:
-        y = norm_r0_plus_norm_r + A * (psi * c3(psi) - 1) / c2(psi) ** 0.5
         if A > 0.0:
             # Readjust xi_low until y > 0.0
             # Translated directly from Vallado
-            while y < 0.0:
-                psi_low = psi
+            ynegktr = 1
+            while y < 0.0 and ynegktr < 10:
                 psi = (
                     0.8
                     * (1.0 / c3(psi))
                     * (1.0 - norm_r0_times_norm_r * np.sqrt(c2(psi)) / A)
                 )
-                y = norm_r0_plus_norm_r + A * (psi * c3(psi) - 1) / c2(psi) ** 0.5
-
-        xi = np.sqrt(y / c2(psi))
-        tof_new = (xi ** 3 * c3(psi) + A * np.sqrt(y)) / np.sqrt(k)
-
-        # Convergence check
-        if np.abs((tof_new - tof) / tof) < rtol:
-            break
-        count += 1
-        # Bisection check
-        condition = tof_new <= tof
-        psi_low = psi_low + (psi - psi_low) * condition
-        psi_up = psi_up + (psi - psi_up) * (not condition)
+                psi_low = psi
+                if abs(c2(psi)) > rtol:
+                    y = norm_r0_plus_norm_r + A * (
+                        (1.0 - psi * c3(psi)) / c2(psi) ** 0.5
+                    )
+                else:
+                    y = norm_r0_plus_norm_r
+
+                ynegktr += 1
+
+        if ynegktr < 10:
+            if abs(c2(psi)) > rtol:
+                xi = np.sqrt(y / c2(psi))
+            else:
+                xi = 0.0
+
+            tof_new = (xi ** 3 * c3(psi) + A * np.sqrt(y)) / np.sqrt(k)
+
+            # Convergence check
+            if np.abs((tof_new - tof) / tof) < rtol:
+                break
+
+            psi_new = _newton_rhapson(psi, A, xi, y, k, tof, tof_new)
+            psi_new, psi_low, psi_up = _check_bounds(
+                psi, psi_new, psi_up, psi_low, tof_new, tof
+            )
+            psi = psi_new
+            count += 1
+
+            # make sure the first guess isn't too close
+            if abs(tof_new - tof) < rtol and count == 1:
+                tof_new = tof - 1.0
```

Comment:
Am I missing an `else: break` if `ynegktr < 10` is False? To reach the `RuntimeError` below?

Reply:
What I'm surprised is, there is already
 ```python
    while count < numiter and ynegktr <= 10:
```
so, why again ```ynegktr < 10```. I'm thinking it's probably because of the first guess not being too close.

Reply:
**FIXED:**
```python
            # Convergence check
            if converged:
                # make sure the first guess isn't too close
                if count == 1:
                    tof_new = tof - 1.0 
                else:
                    break
```
where the ```break``` statement would land us to the RuntimeError :D